doc: matter: Add documentation for Matter Custom Cluster Editor. (#21417)
nrfconnect/sdk-nrf: doc/nrf/protocols/matter/overview/manufacturer_cluster_editor_tool.rst

Use the system file dialog to select the XML cluster file.
If the file contains more than one cluster, the tool shows the list of available clusters in the context menu.

For example:

.. figure:: images/matter_cluster_tool/matter_cluster_tool_load_multiple_clusters.png
   :alt: Load cluster file
   :align: center

Select the cluster from the list to load a specific cluster definition or cluster extension.
After loading the cluster file, all tabs are populated with the data from the loaded cluster.

Saving the cluster file
-----------------------

To save the cluster file, click the :guilabel:`Save cluster to file` button.
Use the system file dialog to select the path and file name to save the cluster.

Saving the cluster extension file
---------------------------------

To save the cluster extension file, click the :guilabel:`Save extension to file` button.
Use the system file dialog to select the path and file name to save the cluster extension.

The tool checks the difference between the current cluster and the loaded cluster extension and saves the difference to the XML file.

> If there is no difference, the tool shows a message indicating that there is no data to create a cluster extension, see figure below:

Reviewer: ```suggestion
If there is no difference, the tool shows a message indicating that there is no data to create a cluster extension, see the figure:
```

You can add, edit or delete the elements in the table using the edit box interface after clicking the :guilabel:`Edit` button in the table.

Edit box
--------

The edit box is a dialog window that allows you to edit a row.
Click the pencil icon on each row of the table or the :guilabel:`Add` button in the top of the table to start editing.

The following figure shows location of the pencil icon and the :guilabel:`Add` button:

.. figure:: images/matter_cluster_tool/matter_cluster_tool_start_edit.png
   :alt: Start edit

To remove a row from the table, click the trash can icon located under each element in the list.

The following figure shows an example of the trash icon location:

.. figure:: images/matter_cluster_tool/matter_cluster_tool_remove_row.png
   :alt: Remove row

The contents of the edit box depends on the type of the tab, but in general it consists of the following elements:

* Fields to be filled with the specific information about the element.
* Additional buttons to add element-specific nested information.

> For example, the :guilabel:`Access` button in the edit box dialog window of the :guilabel:`Attributes` tab allows to add access control information for the attribute, see figure below:

Reviewer: ```suggestion
For example, the :guilabel:`Access` button in the edit box dialog window of the :guilabel:`Attributes` tab allows to add access control information for the attribute, see the figure:
```

.. figure:: images/matter_cluster_tool/matter_cluster_tool_edit_box.png
   :alt: Edit box

You must fill in all required fields marked with the asterisk before saving the element.

> If any of the required fields is not filled, you will see an error message when trying to save the element, see figure below:

Reviewer: ```suggestion
If any of the required fields is not filled, you will see an error message when trying to save the element, see the figure:
```

.. figure:: images/matter_cluster_tool/matter_cluster_tool_edit_box_error.png
   :alt: Edit box error

After clicking an additional button in the edit box, you will see a new dialog with the list of elements.

To add a new element, click the plus icon in the top of the list.
To remove an element, click the trash can icon located under each element in the list.

> For example, the :guilabel:`Access` button in the edit box dialog window of the :guilabel:`Attributes` tab allows to add access control information for the attribute, see figure below:

Reviewer: ```suggestion
For example, the :guilabel:`Access` button in the edit box dialog window of the :guilabel:`Attributes` tab allows to add access control information for the attribute, see the figure:
```

.. figure:: images/matter_cluster_tool/matter_cluster_tool_edit_box_access.png
   :alt: Edit box access

> The number visible on the right upper corner of the additional button represents the current number of elements in the list, see figure below:

Reviewer: ```suggestion
The number visible on the right upper corner of the additional button represents the current number of elements in the list, see the figure:
```

.. figure:: images/matter_cluster_tool/matter_cluster_tool_edit_box_access_number.png
   :alt: Edit box access number

Details
-------

You can expand each row in the table to see the details of the element.
Each tab has a dedicated details implementation, so the content depends on the type of the element.

> To see the details of an element, click the arrow icon on the left side of the row, see figure below:

Reviewer: ```suggestion
To see the details of an element, click the arrow icon on the left side of the row, see the figure:
```

.. figure:: images/matter_cluster_tool/matter_cluster_tool_details_button.png
   :alt: Details button

The following elements can be displayed in the details depending on the type of the element and its content:

* Various text fields with a specific information about the element.

  For example:

  .. figure:: images/matter_cluster_tool/matter_cluster_tool_edit_box_details_text_field.png
     :alt: Edit box details text field

* List of additional elements.
  Click the button in this field to open a dedicated dialog with the list of elements.

>   The figure below shows an example of details with a list of elements:

Reviewer: ```suggestion
  The following figure shows an example of details with a list of elements:
```

#. Click the :guilabel:`Save extension to file` button to save the cluster extension to a file.

.. _ug_nrfconnect_manufacturer_cluster_editor_tool_creating_new_device_type:

Creating a new device type
**************************

This section describes how you can create a new device type and save it as an XML file using the tool.
To create a new device type, complete the following steps:

1. Open the :guilabel:`Device type` tab and fill in the required fields marked with the asterisk.
   You can fill also the optional fields.
   To see tooltips for the fields, hover your mouse over them.

#. Click the :guilabel:`Add Cluster assignment to device type` button to add the required clusters to the device type.
   You will see the dialog window with the all fields to be filled in.

   For example:

   .. figure:: images/matter_cluster_tool/matter_cluster_tool_device_type_add_cluster.png
      :alt: Add cluster assignment to device type

#. Click the buttons in the dialog window to assign required Attributes, Commands and Events by the cluster.

>    To assign a new element, click the :guilabel:`plus` icon button and write the exact name of the each element.

Reviewer: ```suggestion
   To assign a new element, click the plus icon and write the exact name of the each element.
```

   For example:

   .. figure:: images/matter_cluster_tool/matter_cluster_tool_device_type_add_cluster_dialog.png
      :alt: Add cluster assignment to device type dialog

#. You can also specify the features of the device type by clicking the :guilabel:`Features` button.

>    In the new dialog window, click the :guilabel:`plus` icon button to add a new feature, and fill in the code and name of the feature.

Reviewer: ```suggestion
   In the new dialog window, click the plus icon to add a new feature, and fill in the code and name of the feature.
```

        .. group-tab:: Linux

            .. code-block:: console

               a. Make the file executable:

                  .. code-block:: console

                     chmod +x ./nrfconnect-matter-manufacturer-cluster-editor.Appimage

               b. Click twice on the :file:`nrfconnect-matter-manufacturer-cluster-editor.Appimage` file, or use the following command:

                  .. code-block:: console

                     ./nrfconnect-matter-manufacturer-cluster-editor.Appimage

.. _ug_nrfconnect_manufacturer_cluster_editor_tool_basic_functionalities:

Basic functionalities
*********************

The tool UI is created based on the shared commodities for developing `nRF Connect for Desktop`_ applications.
The following sections describe the basic parts of the tool and their functionalities.

Side panel
==========

On the left side of the tool, you can see the navigation panel with the following elements:

* A name bar on the top of the panel with the name of the cluster.
  It represents a name of the cluster, and the name of the file, where the cluster is saved.
  You can change the name of the cluster by clicking on the name bar and typing a new name.

>   The new name shall be unique across all available clusters in the Matter data model.

Author: Hmm, here we must inform users that they shall choose a unique name. Maybe `must` fits better here?

* A :guilabel:`XML cluster file` section that contains two buttons:

  * :guilabel:`Load from file` button, which loads the cluster definition or cluster extension from the XML file.

>   * :guilabel:`Save cluster to file` button, which saves the current cluster definition to the XML file.

Reviewer: ```suggestion
  * :guilabel:`Save cluster to file` that saves the current cluster definition to an XML file.
```

    This button can be used only for the new cluster.

> * A :guilabel:`XML cluster extension file` tab, which contains the button :guilabel:`Save extension to file`, which checks the difference between the current filled-in values and the loaded cluster and saves the difference to the XML file.

Reviewer: ```suggestion
* A :guilabel:`XML cluster extension file` tab with the :guilabel:`Save extension to file` button that checks the difference between the current filled-in values and the loaded cluster and saves the difference to the XML file.
```

.. figure:: images/matter_cluster_tool/matter_cluster_tool_side_panel.png
   :alt: Side panel

You can use the :guilabel:`Show side panel` toggle button to show or hide the side panel.

Loading the cluster file
------------------------

> You can load an XML file that contains the cluster definition or cluster extension by clicking the :guilabel:`Load from file` button.

Reviewer: ```suggestion
To load an XML file that contains the cluster definition or cluster extension, click the :guilabel:`Load from file` button.
```

Use the system file dialog to select the XML cluster file.
If the file contains more than one cluster, the tool shows the list of available clusters in the context menu.

For example:

.. figure:: images/matter_cluster_tool/matter_cluster_tool_load_multiple_clusters.png
   :alt: Load cluster file
   :align: center

Select the cluster from the list to load a specific cluster definition or cluster extension.
After loading the cluster file, all tabs are populated with the data from the loaded cluster.

Saving the cluster file
-----------------------

You can save the cluster file to the file using the :guilabel:`Save cluster to file` button.
Use the system file dialog to select the path and file name to save the cluster.

Saving the cluster extension file
---------------------------------

> You can save the cluster extension file to the file using the :guilabel:`Save extension to file` button.

Reviewer: ```suggestion
To save the cluster extension file, click the :guilabel:`Save extension to file` button.
```

Use the system file dialog to select the path and file name to save the cluster extension.

The tool checks the difference between the current cluster and the loaded cluster extension and saves the difference to the XML file.

> If there is no difference, the tool shows the message that there is no data to create a cluster extension.

Reviewer: ```suggestion
If there is no difference, the tool shows a message indicating that there is no data to create a cluster extension.
```

.. figure:: images/matter_cluster_tool/matter_cluster_tool_save_extension_no_difference.png
   :alt: Save extension no difference

Main panel
==========

> The main panel is the center part of the tool and contains the following tabs, each representing a different element of an XML cluster file:

Reviewer: ```suggestion
The main panel is the central part of the tool and contains the following tabs, each representing a different element of an XML cluster file:
```

* :guilabel:`Cluster` - Contains the cluster definition, description, and code
* :guilabel:`Commands` - Lists all commands assigned to the cluster
* :guilabel:`Attributes` - Shows all attributes associated with the cluster
* :guilabel:`Events` - Displays all events defined for the cluster
* :guilabel:`Enums` - Contains all enumeration types used by the cluster
* :guilabel:`Structs` - Lists all structure types defined for the cluster
* :guilabel:`Device type` - Provides the device type definition

.. figure:: images/matter_cluster_tool/matter_cluster_tool_main_panel.png
   :alt: Main panel

> The :guilabel:`Cluster` and :guilabel:`Device type` tabs consist of fields that can be edited directly in the opened window.

Reviewer: ```suggestion
The :guilabel:`Cluster` and :guilabel:`Device type` tabs consist of fields that you can edit directly in the opened window.
```

The :guilabel:`Commands`, :guilabel:`Attributes`, :guilabel:`Events`, :guilabel:`Enums`, and :guilabel:`Structs` tabs consist of the table with the list of elements.
You can add, edit or delete the elements in the table using the edit box interface after clicking on the edit button in the table.

Edit box
--------

The edit box is a dialog window that allows you to edit a row.

> Click the :guilabel:`Pencil` icon button on each row of the table or :guilabel:`Add` button in the top of the table to start editing.

Reviewer: ```suggestion
Click the pencil icon on each row of the table or the :guilabel:`Add` button in the top of the table to start editing.
```

.. figure:: images/matter_cluster_tool/matter_cluster_tool_start_edit.png
   :alt: Start edit

> To remove a row from the table, click the :guilabel:`Trash` icon button which is located under each element in the list.

Reviewer: ```suggestion
To remove a row from the table, click the trash can icon located under each element in the list.
```

.. figure:: images/matter_cluster_tool/matter_cluster_tool_remove_row.png
   :alt: Remove row

> The edit box content depends on the type of the page, but in general it consists of the following elements:

Reviewer: ```suggestion
The contents of the edit box depends on the type of the tab, but in general it consists of the following elements:
```

* Fields to be filled with the specific information about the element.

> * Additional buttons to add a element-specific nested information.

Reviewer: ```suggestion
* Additional buttons to add element-specific nested information.
```

>   For example, the :guilabel:`Access` button in the edit box dialog window of the :guilabel:`Attributes` tab allows to add a access control information for the attribute.

Reviewer: ```suggestion
  For example, the :guilabel:`Access` button in the edit box dialog window of the :guilabel:`Attributes` tab allows to add access control information for the attribute.
```

.. figure:: images/matter_cluster_tool/matter_cluster_tool_edit_box.png
   :alt: Edit box

You must fill in all required fields marked with the asterisk before saving the element.

> If at least one of the required fields is not filled, you will see the error message while trying to save the element.

Reviewer: ```suggestion
If any of the required fields is not filled, you will see an error message when trying to save the element.
```

.. figure:: images/matter_cluster_tool/matter_cluster_tool_edit_box_error.png
   :alt: Edit box error

After clicking an additional button in the edit box, you will see a new dialog with the list of elements.

> To add a new element, click the :guilabel:`Add` icon button in the top of the list.

Reviewer: ```suggestion
To add a new element, click the plus icon in the top of the list.
```

> To remove an element, click the :guilabel:`Trash` icon button which is located under each element in the list.

Reviewer: ```suggestion
To remove an element, click the trash can icon located under each element in the list.
```

> For example, the :guilabel:`Access` button in the edit box dialog window of the :guilabel:`Attributes` tab allows to add a access control information for the attribute.

Reviewer: ```suggestion
For example, the :guilabel:`Access` button in the edit box dialog window of the :guilabel:`Attributes` tab allows to add access control information for the attribute.
```

.. figure:: images/matter_cluster_tool/matter_cluster_tool_edit_box_access.png
   :alt: Edit box access

The number visible on the right upper corner of the additional button represents the current number of elements in the list.

.. figure:: images/matter_cluster_tool/matter_cluster_tool_edit_box_access_number.png
   :alt: Edit box access number

Details
-------

> Each row in the table can be expanded to see the details of the element.

Reviewer: ```suggestion
You can expand each row in the table to see the details of the element.
```

> Each tab has a dedicated details implementation, so the details content depends on the type of the element.

Reviewer: ```suggestion
Each tab has a dedicated details implementation, so the content depends on the type of the element.
```

> To see the details of the element, click the :guilabel:`arrow` icon button in the left side of the row.

Reviewer: ```suggestion
To see the details of an element, click the arrow icon on the left side of the row.
```

.. figure:: images/matter_cluster_tool/matter_cluster_tool_details_button.png
   :alt: Details button

The following elements can be displayed in the details depending on the type of the element and its content:

* Various text fields with a specific information about the element.

  For example:

  .. figure:: images/matter_cluster_tool/matter_cluster_tool_edit_box_details_text_field.png
     :alt: Edit box details text field

* List of additional elements.

>   You can click the button in this field to open a dedicated dialog with the list of elements.

Reviewer: ```suggestion
  Click the button in this field to open a dedicated dialog with the list of elements.
```

  For example:

  .. figure:: images/matter_cluster_tool/matter_cluster_tool_edit_box_details_list.png
     :alt: Edit box details list

  .. figure:: images/matter_cluster_tool/matter_cluster_tool_edit_box_details_list_dialog.png
     :alt: Edit box details list dialog

* List of boolean elements set to true.

  For example, nullable, reportable and optional elements are set to true:

  .. figure:: images/matter_cluster_tool/matter_cluster_tool_edit_box_details_boolean.png
     :alt: Edit box details boolean

* List of IDs of the clusters to which the element belongs.

  For example:

  .. figure:: images/matter_cluster_tool/matter_cluster_tool_edit_box_details_clusters.png
     :alt: Edit box details clusters

.. _ug_nrfconnect_manufacturer_cluster_editor_tool_creating_new_cluster:

Creating a new Matter cluster
*****************************

> In this section, you can learn how to create a new Matter cluster using the Manufacturer Cluster Editor tool and save it as an XML file.

Reviewer: ```suggestion
This section describes how you can create a new Matter cluster and save it as an XML file using the tool.
```

> To create a new cluster definition, follow these steps:

Reviewer: ```suggestion
To create a new cluster definition, complete the following steps:
```

> 1. Navigate to the :guilabel:`Cluster` tab and fill in the required fields marked with the asterisk.

Reviewer: ```suggestion
1. Open the :guilabel:`Cluster` tab and fill in the required fields marked with the asterisk.
```

>    Optionally, you fill in the optional fields.

Reviewer: ```suggestion
   You can fill also the optional fields.
```

>    To see tooltips for the fields, hoover your mouse over the field.

Reviewer: ```suggestion
   To see tooltips for the fields, hover your mouse over them.
```

> #. Navigate one by one to the :guilabel:`Commands`, :guilabel:`Attributes`, and :guilabel:`Events` tabs and add the required elements to the cluster.

Reviewer: ```suggestion
#. Open one by one the :guilabel:`Commands`, :guilabel:`Attributes`, and :guilabel:`Events` tabs and add the required elements to the cluster.
```

>    Use the :guilabel:`Add` button in the top of each element tab to add a new element to the cluster.

Reviewer: ```suggestion
   Click the :guilabel:`Add` button on the top of each element tab to add a new element to the cluster.
```

   See the :ref:`ug_nrfconnect_manufacturer_cluster_editor_tool_basic_functionalities` section to learn how to add a new element to the cluster.

#. Navigate one by one to the :guilabel:`Enums` and :guilabel:`Structs` tabs and add the required elements.

>    The enums and structs does not belong to the cluster, but they can be assigned to one or more clusters.

Reviewer: ```suggestion
   The enums and structs do not belong to the cluster, but you can assign them to one or more clusters.
```

   See the :ref:`ug_nrfconnect_manufacturer_cluster_editor_tool_basic_functionalities` section to learn how to add a new element to the cluster.

> #. Save the cluster to the file using the :guilabel:`Save cluster to file` button.

Reviewer: ```suggestion
#. Click :guilabel:`Save` to save the cluster to a file.
```

.. _ug_nrfconnect_manufacturer_cluster_editor_tool_creating_cluster_extension:

Creating a cluster extension
****************************

> In this section, you can learn how to create a cluster extension using the Manufacturer Cluster Editor tool and save it as an XML file.

Reviewer: ```suggestion
This section describes how you can create a cluster extension and save it as an XML file using the tool.
```

> To create a cluster extension, follow these steps:

Reviewer: ```suggestion
To create a cluster extension, complete the following steps:
```

> 1. Load an existing cluster definition using the :guilabel:`Load from file` button.

Reviewer: ```suggestion
1. Click the :guilabel:`Load from file` button to load an existing cluster definition.
```

   Use the system file dialog to select the XML cluster file.
   If the file contains more than one cluster, the tool shows the list of available clusters in the context menu.

> #. Navigate one by one to the :guilabel:`Commands`, :guilabel:`Attributes`, and :guilabel:`Events` tabs and add the required elements to the cluster.

Reviewer: ```suggestion
#. Open one by one the :guilabel:`Commands`, :guilabel:`Attributes`, and :guilabel:`Events` tabs and add the required elements to the cluster.
```

   Use the :guilabel:`Add` button in the top of each element tab to add a new element to the cluster.
   See the :ref:`ug_nrfconnect_manufacturer_cluster_editor_tool_basic_functionalities` section to learn how to add a new element to the cluster.

> #. Navigate one by one to the :guilabel:`Enums` and :guilabel:`Structs` tabs and add the required elements.

Reviewer: ```suggestion
#. Open one by one the :guilabel:`Enums` and :guilabel:`Structs` tabs and add the required elements.
```

Reviewer: And do the same tricks to the rest.

            .. code-block:: console

                ./nrfconnect-matter-manufacturer-cluster-editor.app/Contents/MacOS/nrfconnect-matter-manufacturer-cluster-editor

        .. group-tab:: macOS (ARM64)

            .. code-block:: console

                ./nrfconnect-matter-manufacturer-cluster-editor.app/Contents/MacOS/nrfconnect-matter-manufacturer-cluster-editor

        .. group-tab:: Linux

            .. code-block:: console

               a. Make the file executable:

                  .. code-block:: console

                     chmod +x ./nrfconnect-matter-manufacturer-cluster-editor.Appimage

               b. Click twice on the :file:`nrfconnect-matter-manufacturer-cluster-editor.Appimage` file, or use the following command:

                  .. code-block:: console

                     ./nrfconnect-matter-manufacturer-cluster-editor.Appimage

.. _ug_nrfconnect_manufacturer_cluster_editor_tool_basic_functionalities:

Basic functionalities
*********************

The tool UI is created based on the shared commodities for developing `nRF Connect for Desktop`_ applications.

> If you are familiar with them, you will feel comfortable using the tool.

Author: You will feel anxious then. I will change it. :)
